server/src/main/java/org/opensearch/node/Node.java

```diff
@@ -1155,6 +1134,35 @@ protected Node(
                 taskHeaders,
                 tracer
             );
+            final QueryGroupService queryGroupService = new QueryGroupService(transportService.getLocalNode(), clusterService); // We will
+                                                                                                                                // need to
+                                                                                                                                // replace
+                                                                                                                                // this with
+                                                                                                                                // actual
+            // instance of the
+            // queryGroupService
+            queryGroupServiceSetOnce.set(queryGroupService);
+
+            final QueryGroupRequestOperationListener queryGroupRequestOperationListener = new QueryGroupRequestOperationListener(
+                queryGroupService,
+                threadPool
+            );
+
+            // register all standard SearchRequestOperationsCompositeListenerFactory to the SearchRequestOperationsCompositeListenerFactory
+            final SearchRequestOperationsCompositeListenerFactory searchRequestOperationsCompositeListenerFactory =
+                new SearchRequestOperationsCompositeListenerFactory(
+                    Stream.concat(
+                        Stream.of(
+                            searchRequestStats,
+                            searchRequestSlowLog,
+                            searchTaskRequestOperationsListener,
+                            queryGroupRequestOperationListener
+                        ),
+                        pluginComponents.stream()
+                            .filter(p -> p instanceof SearchRequestOperationsListener)
+                            .map(p -> (SearchRequestOperationsListener) p)
+                    ).toArray(SearchRequestOperationsListener[]::new)
+                );
```

Comment:
Why not initialize the `QueryGroupRequestOperationListener` and `QueryGroupService` before the original `SearchRequestOperationsCompositeListenerFactory` creation, instead of moving all the logic here?

Reply:
The creation of `QueryGroupService` needs `transportService`, which is only initialized in line 1127. So we have to move `SearchRequestOperationsCompositeListenerFactory` down

Reply:
We don't need `TransportService` into `QueryGroupService`. The `nodeStats` within `QueryGroupService` should be completely oblivious of the `localNodeId`. It should just return the `new QueryGroupStats(statsHolderMap)`. TransportAction can create the `WlmStats` object from `QueryGroupStats`.